### PR TITLE
use threading to fix time timeout problem

### PR DESCRIPTION
**Pull request by karta0807913**
1. Set timeout to the target function
2. Reduce the impact of multiple queries

### CODE REVIEW

**Review by laixintao — CHANGES_REQUESTED** [review]
It's generally a very good implementation.

Actually there is already an [implementation](https://github.com/python/cpython/blob/3fc5d84046ddbd66abac5b598956ea34605a4e5d/Lib/functools.py#L525) based on double linked list, considered all the details, and it's usage is very simple: https://docs.python.org/3/library/functools.html#functools.lru_cache , I prefer to use this one, the only issue is that, it's not "timed", only evict items when the `maxsize` reached, which is not good for our service discovery case. Some solutions [here](https://stackoverflow.com/questions/31771286/python-in-memory-cache-with-time-to-live).

The `threading.join()` implementation for timeout is excellent. But I saw that we only simply check if the thread is finished or not, without killing it. I don't have any ideas on that, maybe killing a thread is generally a bad idea, so maybe we only give a lock to every script so that even the thread is not terminated we only have one thread running per script? Python's built in `lru_cache` also didn't lock functions, means there could be same exactly function running at the same time.

File: test/test_timeout/test_timeout.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    # test if the decorator can raise an exception after timeout.
+    with pytest.raises(TimeoutException):
+        _ = havy_function()
+
+    # test if the decorator can cache the result.
+    time.sleep(2)
+    first_call = havy_function()
+    second_call = havy_function()
+    assert first_call == second_call, "the function did't cache the result :("
```
[the remark this answers is not in the document]
Comment (laixintao) [reply test/test_timeout/test_timeout.py:22]: so can use `object()` to replace `random`

File: prometheus_http_sd/sd.py [hunk 2]
```diff
@@ -143,6 +145,75 @@ def run_json(file_path: str) -> TargetList:
...
+            }
+
+            def target_function():
+                try:
+                    cache["response"] = function(*arg, **kwargs)
+                    cache["timestamp"] = time.time()
+                except Exception as e:
+                    cache["error"] = e
+
+            with lock:
```
[the remark this answers is not in the document]
Comment (karta0807913) [reply prometheus_http_sd/sd.py:188]: yes, but I only do a little job in the lock, so I think it won't affect so much.

File: prometheus_http_sd/sd.py [hunk 3]
```diff
@@ -143,6 +145,75 @@ def run_json(file_path: str) -> TargetList:
...
+                if key in thread_cache:
+                    cache_expire_time = (
+                        thread_cache[key]["timestamp"] + cache_time
+                    )
+                    current_time = time.time()
+                    if (
+                        thread_cache[key]["thread"].is_alive()
+                        or cache_expire_time > current_time
+                    ):
+                        cache = thread_cache[key]
```
[the remark this answers is not in the document]
Comment (karta0807913) [reply prometheus_http_sd/sd.py:199]: make sense. let me use another thread to clean them.

**Comment by karta0807913** [comment]
I refrained from utilizing the `lru_cache` due to the potential variability of the target generator over time. If a URL is repeatedly called, the cache will not expire, causing the targets to remain static. This would not be beneficial for our monitoring virtual machines. Also, the `lru_cache` function is not thread-safe. Therefore, I prefer using the time-based cache to handle the cache.

```
import time
import functools
import threading

@functools.lru_cache
def cache_function():
    time.sleep(10)
    print("HI")
    return 2

threading.Thread(target=cache_function).start()
threading.Thread(target=cache_function).start()
threading.Thread(target=cache_function).start()
```

output is
```
HI
HI
HI
```

Therefore, I used a lock inside the decorator to use the same thread for the cache -- this won't affect the users so much :D
Users still can call the function parallelly, because the "thread.wait()" is outside of the lock scope. The lock scope only controls the cache usage.

About the threading problem, Python not support stop the thread during the thread is executing (as same as another languages).
Also, not stopping the thread allow users can get the result after serval function calls. For example:
```
First Function call (timeout)
└─┘
Second Function call (timeout)
             └─┘
Third Function call(get result)
                       └─┘
Function Operating   Cache time
└───────────────────┴───────────┘
```

And I agree we should implement the garbage collection for this function. Let me implement it by using the heap :D

File: test/test_timeout/test_timeout.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (karta0807913) [reply test/test_timeout/test_timeout.py]: done

File: prometheus_http_sd/sd.py [hunk 4]
```diff
@@ -143,6 +145,75 @@ def run_json(file_path: str) -> TargetList:
...
+                if cache["thread"] is None:
+                    cache["thread"] = threading.Thread(target=target_function)
+                    cache["thread"].start()
+                thread_cache[key] = cache
+            cache["thread"].join(timeout)
+            if cache["thread"].is_alive():
+                raise TimeoutException("target function timeout!")
+            if cache["error"] is not None:
```
[the remark this answers is not in the document]
Comment (karta0807913) [reply prometheus_http_sd/sd.py]: done :D

File: prometheus_http_sd/sd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karta0807913) [reply prometheus_http_sd/sd.py]: done, please help to review :D

File: prometheus_http_sd/decroator.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        )
+
+    def _cache_garbage_collection(self):
+        def can_iterate():
+            with self.heap_lock:
+                if len(self.heap) == 0 or self.heap[0][0] > time.time():
+                    return False
+            return True
+
+        print(self.heap, can_iterate())
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py:70]: use logging module please

**Review by laixintao — CHANGES_REQUESTED** [review]
Can you add metrics to those:

- last gc timestamp
- gc time count
- current cached item count

So we can monitor http-sd itself

**Comment by karta0807913** [comment]
sure, let me add the metrics

File: prometheus_http_sd/decroator.py (hunk 5 above)
Comment (karta0807913) [reply prometheus_http_sd/decroator.py:70]: oh, I forgot to delete the test code LOL

File: prometheus_http_sd/decroator.py [hunk 6]
```diff
@@ -82,6 +88,10 @@ def can_iterate():
                 if self.is_expired(self.thread_cache[_key]):
                     del self.thread_cache[_key]
         self.garbage_collection_timestamp = time.time()
+        logger.Info(f"""Finish garbage collection finish metrics:
+        Finish garbage collection time: {self.garbage_collection_timestamp},
+        len(heap): {len(self.heap)}
+        len(thread_cache): {self.thread_cache}""")
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py]: sorry, I didn't make myself clear, I mean, expose prometheus metrics, like this:

https://github.com/laixintao/prometheus-http-sd/blob/778dbbba5508dff96d7ff602865812574d3d8570/prometheus_http_sd/sd.py#L24

File: prometheus_http_sd/decroator.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+)
+
+_heap_cache_count = Gauge(
+    "garbage_collection_heap_count",
+    "Show current heap length",
+    ["name"],
+)
+
+_collection_run_interval = Histogram(
+    "garbage_collection_run_interval",
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py:26]:
```suggestion
     "http_sd_garbage_collection_run_interval_seconds_bucket",
```

there is a naming convention: https://prometheus.io/docs/practices/naming/

File: prometheus_http_sd/decroator.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
+import time
+import heapq
+import threading
+
+from prometheus_client import Gauge, Counter, Histogram
+
+_collected_total = Counter(
+    "garbage_collection_collected_items",
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py:8]:
```suggestion
    "httpsd_garbage_collection_collected_items_total",
```

File: prometheus_http_sd/decroator.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    "The total count of the garbage collection collected items.",
+    ["name"],
+)
+
+_thread_cache_count = Gauge(
+    "garbage_collection_cache_count",
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py:14]:
```suggestion
    "httpsd_garbage_collection_cache_count",
```

File: prometheus_http_sd/decroator.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    "Show current thread_cache count",
+    ["name"],
+)
+
+_heap_cache_count = Gauge(
+    "garbage_collection_heap_count",
```
Comment (laixintao) [inline prometheus_http_sd/decroator.py:20]:
```suggestion
    "httpsd_garbage_collection_heap_count",
```

**Review by laixintao — CHANGES_REQUESTED** [review]

**Pull request merged**